consensus/integration_test_types.go

```diff
@@ -0,0 +1,443 @@
+				node.handleVoteQCConstructionEvent(event)
+				break
+			case FastHotStuffEventTypeConstructTimeoutQC:
+				node.handleTimeoutQCConstructionEvent(event)
+				break
+			}
+		case <-node.quit:
+			return
+		}
+	}
+}
+
+func (node *validatorNode) handleVoteEvent(event *FastHotStuffEvent) {
+	payload := GetVoteSignaturePayload(event.View, event.TipBlockHash)
+	signature, err := node.privateKey.Sign(payload[:])
+	if err != nil {
+		panic(err)
+	}
+
+	vote := &voteMessage{
+		view:      event.View,
+		blockHash: event.TipBlockHash,
+		publicKey: node.privateKey.PublicKey(),
+		signature: signature,
+	}
+
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessVote(vote)
+	}
+}
+
+func (node *validatorNode) handleTimeoutEvent(event *FastHotStuffEvent) {
+	node.lock.Lock()
+	defer node.lock.Unlock()
+
+	// Skip if the timed out view from the event is stale
+	if node.eventLoop.currentView != event.View {
+		return
+	}
+
+	// Skip if the node can't advance to the next view for some reason.
+	// This should never happen.
+	if _, err := node.eventLoop.AdvanceViewOnTimeout(); err == nil {
+		node.broadcastTimeout(event)
+	}
+}
+
+func (node *validatorNode) broadcastTimeout(event *FastHotStuffEvent) {
+	payload := GetTimeoutSignaturePayload(event.View, event.QC.GetView())
+	signature, err := node.privateKey.Sign(payload[:])
+	if err != nil {
+		panic(err)
+	}
+
+	timeout := &timeoutMessage{
+		view:      event.View,
+		highQC:    event.QC,
+		publicKey: node.privateKey.PublicKey(),
+		signature: signature,
+	}
+
+	// Broadcast the block to all validators.
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessTimeout(timeout)
+	}
+}
+
+func (node *validatorNode) handleVoteQCConstructionEvent(event *FastHotStuffEvent) {
+	if !node.isBlockProposer {
+		return
+	}
+
+	// Skip if the view from the event is stale. This can happen if the node
+	// has advances to the next view before the block is processed.
+	if node.eventLoop.currentView != event.View {
+		return
+	}
+
+	block := &block{
+		view:      event.View,
+		blockHash: createDummyBlockHash(),
+		height:    event.TipBlockHeight + 1,
+		qc:        event.QC,
+	}
+
+	// Broadcast the block to all validators.
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessBlock(block)
+	}
+}
+
+func (node *validatorNode) handleTimeoutQCConstructionEvent(event *FastHotStuffEvent) {
+	if !node.isBlockProposer {
+		return
+	}
+
+	// Skip if the view from the event is stale. This can happen if the node
+	// has advances to the next view before the block is processed.
+	if node.eventLoop.currentView != event.View {
+		return
+	}
+
+	block := &block{
+		view:        event.View,
+		blockHash:   createDummyBlockHash(),
+		height:      event.TipBlockHeight + 1,
+		aggregateQC: event.AggregateQC,
+	}
+
+	// Broadcast the block to all validators.
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessBlock(block)
```

Comment:
Same feedback as for handleVoteQCConstructionEvent. If you're the node proposing the block, you need to process it before this function returns so that you can guarantee that the view has incremented. Not doing this will cause staleness checks to fail erroneously.

Reply:
Same comment as above. All handle.*() functions hold a lock on the event loop when processing incoming messages / events. Once a handler starts running, it is guaranteed to complete before the next one runs